### PR TITLE
[Wasm-GC] Implement packed types in arrays

### PR DESCRIPTION
**Pull request by takikawa**
#### 04de71ca40b4832b65e5c6861c399d91735c309c
<pre>
[Wasm-GC] Implement packed types in arrays
<a href="https://bugs.webkit.org/show_bug.cgi?id=247576">https://bugs.webkit.org/show_bug.cgi?id=247576</a>

Reviewed by Justin Michaud.

This patch implements support for packed types (i8, i16) in Wasm GC
arrays. Packed types are represented with new entries in wasm.json and
are only allowed for use in StorageTypes, which are a new kind of type
used for struct and array type definition fields.

Packed structs are not yet allowed with this patch.

* JSTests/wasm/gc/arrays.js:
(testArrayDeclaration):
* JSTests/wasm/gc/packed-arrays.js: Added.
(module):
(check):
(testArrayGetPacked):
(testArrayGetUWithNewCanonPacked):
(testArrayGetSWithNewCanonPacked):
(testTypeMismatch64):
(testTypeMismatchArrayGet):
(testPackedTypeOutOfContext):
(testSetGetTruncate):
(testArraySet):
(testArrayGetUnreachable):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addArrayNewDefault):
(JSC::Wasm::AirIRGenerator::addArrayGet):
(JSC::Wasm::AirIRGenerator::addStructGet):
(JSC::Wasm::AirIRGenerator::addStructSet):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitStructSet):
(JSC::Wasm::B3IRGenerator::pushArrayNew):
(JSC::Wasm::B3IRGenerator::addArrayNew):
(JSC::Wasm::B3IRGenerator::addArrayNewDefault):
(JSC::Wasm::B3IRGenerator::addArrayGet):
(JSC::Wasm::B3IRGenerator::addStructGet):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isRefType):
(JSC::Wasm::isRefWithRecursiveReference):
(JSC::Wasm::isSubtype):
(JSC::Wasm::isDefaultableType):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addArrayGet):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parsePackedType):
(JSC::Wasm::SectionParser::parseStorageType):
(JSC::Wasm::SectionParser::parseStructType):
(JSC::Wasm::SectionParser::parseArrayType):
* Source/JavaScriptCore/wasm/WasmSectionParser.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
(JSC::LLInt::ArrayGetOperation):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::StructType::dump const):
(JSC::Wasm::StructType::StructType):
(JSC::Wasm::ArrayType::dump const):
(JSC::Wasm::StorageType::dump const):
(JSC::Wasm::computeStructTypeHash):
(JSC::Wasm::computeArrayTypeHash):
(JSC::Wasm::TypeDefinition::replacePlaceholders const):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::StorageType::is const):
(JSC::Wasm::StorageType::as const):
(JSC::Wasm::StorageType::StorageType):
(JSC::Wasm::StorageType::unpacked const):
(JSC::Wasm::StorageType::elementSize const):
(JSC::Wasm::StorageType::operator== const):
(JSC::Wasm::StorageType::operator!= const):
(JSC::Wasm::StorageType::typeCode const):
(JSC::Wasm::StorageType::index const):
(JSC::Wasm::makeString):
(JSC::Wasm::typeSizeInBytes):
* Source/JavaScriptCore/wasm/generateWasm.py:
(Wasm.__init__):
* Source/JavaScriptCore/wasm/generateWasmOpsHeader.py:
(cppMacro):
(cppMacroPacked):
(packedTypeMacroizer):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.cpp:
(JSC::JSWebAssemblyArray::JSWebAssemblyArray):
(JSC::JSWebAssemblyArray::~JSWebAssemblyArray):
(JSC::JSWebAssemblyArray::visitChildrenImpl):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
(JSC::JSWebAssemblyStruct::get const):
(JSC::JSWebAssemblyStruct::set):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/257597@main">https://commits.webkit.org/257597@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/023be503b3251a1c8b99937326a9da2908239f19

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98835 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8048 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31984 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108252 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168510 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8598 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85415 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91369 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106241 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104518 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6513 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90075 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33541 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88335 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21418 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76406 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89588 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1960 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22949 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85383 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1869 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45411 "Found 1 new test failure: media/video-inactive-playback.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28782 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5227 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6825 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42406 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88236 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3266 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19748 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp [hunk 1]
```diff
@@ -3533,11 +3535,14 @@ auto AirIRGenerator::addArrayNewDefault(uint32_t typeIndex, ExpressionType size,
     return { };
 }
 
-auto AirIRGenerator::addArrayGet(uint32_t typeIndex, ExpressionType arrayref, ExpressionType index, ExpressionType& result) -> PartialResult
+auto AirIRGenerator::addArrayGet(GCOpType arrayGetKind, uint32_t typeIndex, ExpressionType arrayref, ExpressionType index, ExpressionType& result) -> PartialResult
 {
+    // arrayGetKind must be one of array.get, array.get_u, or array.get_s
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp]: nit: these comments don't seem very helpful?

File: Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp [hunk 2]
```diff
@@ -3560,9 +3565,11 @@ auto AirIRGenerator::addArrayGet(uint32_t typeIndex, ExpressionType arrayref, Ex
     auto getValue = g64();
     // FIXME: Emit this inline.
     // https://bugs.webkit.org/show_bug.cgi?id=245405
-    emitCCall(&operationWasmArrayGet, getValue, instanceValue(), addConstant(Types::I32, typeIndex), arrayref, index);
+    emitCCall(&(arrayGetKind == GCOpType::ArrayGet ? operationWasmArrayGet :
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp]: Can this be a variable?

Comment (justinmichaud) [reply Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp]: Actually, why can't this just be inline? It just adds some bit math and/or sign extend after, right?

File: Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp [hunk 3]
```diff
@@ -2531,7 +2529,6 @@ auto B3IRGenerator::addArraySet(uint32_t typeIndex, ExpressionType arrayref, Exp
     }
 
     // FIXME: Emit this inline.
-    // https://bugs.webkit.org/show_bug.cgi?id=245405
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp]: why is this removed?

File: Source/JavaScriptCore/wasm/WasmFormat.h [hunk 4]
```diff
@@ -104,6 +104,14 @@ inline bool isRefType(Type type)
     return type.isFuncref() || type.isExternref();
 }
 
+// If this is a type, returns true iff its a ref type; if it's a packed type, returns false
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmFormat.h]: nit: it's or it is

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 5]
```diff
@@ -1119,28 +1126,56 @@ JSC_DEFINE_JIT_OPERATION(operationWasmArrayNew, EncodedJSValue, (Instance* insta
...
     JSWebAssemblyArray* array = nullptr;
-    switch (fieldType.type.kind) {
-    case Wasm::TypeKind::I32:
-    case Wasm::TypeKind::F32: {
+
+    switch (elementSize) {
+    case sizeof(uint8_t): {
+        FixedVector<uint8_t> values(size);
+        for (unsigned i = 0; i < size; i++) {
+            // `encValue` must be an unboxed int (since the typechecker guarantees that its type is i32);
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: Let's add an assert here just in case, to aid the fuzzer

File: Source/JavaScriptCore/wasm/WasmTypeDefinition.h [hunk 6]
```diff
@@ -204,8 +204,108 @@ enum Mutability : uint8_t {
...
+
+    TypeIndex index() const
+    {
+        if (is<Type>())
+            return as<Type>()->index;
+        return 0;
+    }
+    void dump(WTF::PrintStream& out) const;
+
+    // private:
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmTypeDefinition.h]: nit

File: Source/JavaScriptCore/wasm/WasmTypeDefinition.h [hunk 7]
```diff
@@ -204,8 +204,108 @@ enum Mutability : uint8_t {
...
+    {
+        m_storageType = std::variant<Type, PackedType>(t);
+    }
+
+    explicit StorageType(PackedType t)
+    {
+        m_storageType = std::variant<Type, PackedType>(t);
+    }
+
+    // If this is a type, returns itself; if it's a packed type, returns i32
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmTypeDefinition.h]: These kinds of comments aren't very helpful. It would be better to give context about why this is needed, for example explaining that this is the storage type for packed GC types, rather than summarizing what the code does.

File: Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h [hunk 8]
```diff
@@ -86,7 +92,23 @@ class JSWebAssemblyArray final : public JSNonFinalObject {
 
     void set(VM& vm, uint32_t index, EncodedJSValue value)
     {
-        switch (m_elementType.kind) {
+        if (m_elementType.type.is<Wasm::PackedType>()) {
+            // `value` is assumed to be an unboxed int64; truncate it to either 8 or 16 bits
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h]: can we assert that this is valid?

**Review by justinmichaud — CHANGES_REQUESTED** [review]
Overall LGTM, but I think we are moving in the opposite of the intended direction for inlining things. Can you take a look?

File: Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp (hunk 1 above)
Comment (takikawa) [reply Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp]: Deleted this comment and some similar ones.

File: Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp (hunk 2 above)
Comment (takikawa) [reply Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp]: I've now inlined all of the signed/unsigned specific operations.

File: Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp (hunk 3 above)
Comment (takikawa) [reply Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp]: looks like a mistaken deletion, I put it back

File: Source/JavaScriptCore/wasm/WasmOperations.cpp (hunk 5 above)
Comment (takikawa) [reply Source/JavaScriptCore/wasm/WasmOperations.cpp]: Done, added a `ASSERT(encValue <= UINT32_MAX);` here

File: Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h (hunk 8 above)
Comment (takikawa) [reply Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h]: Added a similar assertion as above, `ASSERT(value <= UINT32_MAX);`

**Comment by takikawa** [comment]
> Overall LGTM, but I think we are moving in the opposite of the intended direction for inlining things. Can you take a look?

Thanks for the comments, I think I've addressed all the comments now. 

The largest change was inlining more operations. The main operation for get/set should now be minimally modified, and the packed-type specific operations are done in-line in B3/Air and in extra C++ slow path code for LLInt.

I've also made various smaller cleanup changes to the patch throughout.

File: Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp [hunk 9]
```diff
@@ -2474,12 +2478,28 @@ auto B3IRGenerator::addArrayGet(uint32_t typeIndex, ExpressionType arrayref, Exp
...
         patchpoint->setGenerator([=] (CCallHelpers& jit, const StackmapGenerationParams& params) {
             jit.move(params[1].gpr(), params[0].gpr());
         });
-        result = push(patchpoint);
+        Value* postProcess = patchpoint;
+        switch (arrayGetKind) {
+        case GCOpType::ArrayGet:
+        case GCOpType::ArrayGetU:
+            break;
+        case GCOpType::ArrayGetS: {
```
Comment (takikawa) [inline Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:2504]: Added in-lined signed/unsigned specialization here, rather than relying on multiple engine operations.

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 10]
```diff
@@ -1119,25 +1126,46 @@ JSC_DEFINE_JIT_OPERATION(operationWasmArrayNew, EncodedJSValue, (Instance* insta
...
+
     JSWebAssemblyArray* array = nullptr;
-    switch (fieldType.type.kind) {
-    case Wasm::TypeKind::I32:
-    case Wasm::TypeKind::F32: {
+
+    switch (elementSize) {
+    case sizeof(uint8_t): {
+        // `encValue` must be an unboxed int32 (since the typechecker guarantees that its type is i32); so it's safe to truncate it in the cases below.
+        ASSERT(encValue <= UINT32_MAX);
```
Comment (takikawa) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp:1138]: New assertion added

File: Source/JavaScriptCore/wasm/WasmSlowPaths.cpp [hunk 11]
```diff
@@ -347,21 +347,46 @@ WASM_SLOW_PATH_DECL(array_new_default)
...
+WASM_SLOW_PATH_DECL(array_get)
+{
+    return ArrayGetOperation<WasmArrayGet>(callFrame, pc, instance, [=](auto, auto result) {
+        return result;
+    });
+}
+
+WASM_SLOW_PATH_DECL(array_get_s)
+{
+    auto callback = [=](Wasm::StorageType type, EncodedJSValue value) {
```
Comment (takikawa) [inline Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:439]: The signed array get processing is now done via a callback here (no-op callback in the other cases).

File: Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp [hunk 12]
```diff
@@ -3562,10 +3565,27 @@ auto AirIRGenerator::addArrayGet(uint32_t typeIndex, ExpressionType arrayref, Ex
...
-    switch (elementType.kind) {
+    switch (resultType.kind) {
     case TypeKind::I32:
         result = g32();
         append(Move32, getValue, result);
+        switch (arrayGetKind) {
+        case GCOpType::ArrayGet:
+        case GCOpType::ArrayGetU:
+            break;
+        case GCOpType::ArrayGetS: {
```
Comment (takikawa) [inline Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:3711]: Similar to the B3 changes, the signed get post-processing is emitted in-line here now.

File: Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp [hunk 13]
```diff
@@ -3562,10 +3565,27 @@ auto AirIRGenerator::addArrayGet(uint32_t typeIndex, ExpressionType arrayref, Ex
...
+            size_t elementSize = *elementType.as<PackedType>() == PackedType::I8 ? sizeof(uint8_t) : sizeof(uint16_t);
+            uint8_t bitShift = (sizeof(uint32_t) - elementSize) * 8;
+            auto tmpForShift = g32();
+            append(Move32, Arg::imm(bitShift), tmpForShift);
+            addShift(Types::I32, Lshift32, result, tmpForShift, result);
+            addShift(Types::I32, Rshift32, result, tmpForShift, result);
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:3717]: There really isn't an imm form of Lshift32?

File: Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp [hunk 14]
```diff
@@ -2474,12 +2478,28 @@ auto B3IRGenerator::addArrayGet(uint32_t typeIndex, ExpressionType arrayref, Exp
...
+            size_t elementSize = *elementType.as<PackedType>() == PackedType::I8 ? sizeof(uint8_t) : sizeof(uint16_t);
+            uint8_t bitShift = (sizeof(uint32_t) - elementSize) * 8;
+            Value* shiftLeft = m_currentBlock->appendNew<Value>(m_proc, B3::Shl, origin(), patchpoint, m_currentBlock->appendNew<Const32Value>(m_proc, origin(), bitShift));
+            postProcess = m_currentBlock->appendNew<Value>(m_proc, B3::SShr, origin(), shiftLeft  , m_currentBlock->appendNew<Const32Value>(m_proc, origin(), bitShift));
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp]: nit: spacing

**Review by justinmichaud — APPROVED** [review]
r=me

File: Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp (hunk 13 above)
Comment (takikawa) [reply Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:3717]: At least not with the `addShift` helper AFAIK, which is abstracting over different architectures. It might be possible to generalize `addShift` a bit but it didn't seem worth doing right now.
